10/README.md

```diff
@@ -0,0 +1,88 @@
+
+In this BEP, we discuss a possible path towards an architecture more suitable for BigchainDB 2.0.
+
+In this BEP, we evaluate a rewrite of BigchainDB in another language, specifically Golang. The benefits are multiple: Golang it is a fairly simple language, easy to learn and read; it is shown to be efficient and faster than Python and can compile to a binary file; it is used in several big projects, and it's the language of choice for major blockchain projects: this simplifies code reuse if we need it in the future, and allows us a tighter integration with Tendermint.
+
+Rewriting a system to another language is a risky and possibly endless process. To minimize the risk, an incremental approach is proposed. The approach is incremental in the following sense:
+- It is divided into different subsequent steps.
+- Every step is self contained and outputs a usable system.
+- Every step should be a unit of work reasonably small.
+- It is not a requirement to fully migrate the system, the rewrite can be stopped at any point.
+
+From a bird's-eye perspective, the idea is to replace the Tendermint process BigchainDB depends on, with our own process: `bigchaindb-go` (this is not the final name). The new `bigchaindb-go` will allow a tighter implementation with Tendermint Core. The goal is to incrementally replace the HTTP API endpoints implemented in Python with new ones implemented in `bigchaindb-go`.
+
+This approach is not new. Martin Fowler defined this approach in 2004, in his blog post [Strangler Application][strangler:application]. A [paper has been presented][strangler:paper] the same year, during a conference on extreme programming (XP2004). Benefits of this approach, compared to other approaches like _big bang refactoring_, has been proven valid by many [case][strangler:case-study-1] [studies][strangler:case-study-2].
+
+### Step 1: Posting a transaction
+Posting a transaction to BigchainDB is one of the main primitives of the system.
+
+The goal of _Step 1_ is to expose through `bigchaindb-go` one single endpoint fully compatible with [`POST /api/v1/transactions`][bdb:post-tx], and implement the logic required to satisfy the [ABCI][abci] interface.
+
+The integration between the existing BigchainDB Server code (written in Python) and the new `bigchaindb-go` is described in the following section.
+
+About the existing BigchainDB Server:
+- All the HTTP API endpoints will still be served from the existing code.
+- The endpoint `POST /api/v1/transactions` will be just a _pass through proxy_ to the `bigchaindb-go` implementation.
+- A new endpoint `POST /api/v1/validate` will be implemented, to check the validity of a transaction (this **temporary** endpoint will be used by `bigchaindb-go` to do the actual validation, and will give us more time to actually focus on the experiment).
```

Comment:
Shouldn't this be a `GET /api/v1/validate` endpoint, that is, using an HTTP GET method, since it's just to get information, not to post or store some new information.

Also, I wonder why it should be a temporary endpoint. It might be useful for outside users of the BigchainDB network, to find out of a transaction would be valid if it were POSTed at the instant the request was made.

Reply:
Regarding question 1), I realized why you chose to use the HTTP POST method: it's because GET requests aren't supposed to have a HTTP request body, and some web servers put a limit on how big the URI can be, so if you used GET, there would be nowhere to stuff the transaction! Maybe explain that in the document.

Reply:
> Shouldn't this be a GET /api/v1/validate endpoint, that is, using an HTTP GET method, since it's just to get information, not to post or store some new information.

and

> Regarding question 1), I realized why you chose to use the HTTP POST method: it's because GET requests aren't supposed to have a HTTP request body, and some web servers put a limit on how big the URI can be, so if you used GET, there would be nowhere to stuff the transaction! Maybe explain that in the document.

That's a valid point, but it's *controversial* to have a GET with a body, see:
- https://stackoverflow.com/a/983458/597097
- https://www.elastic.co/guide/en/elasticsearch/guide/current/_empty_search.html

> Also, I wonder why it should be a temporary endpoint. It might be useful for outside users of the BigchainDB network, to find out of a transaction would be valid if it were POSTed at the instant the request was made.

This is true, I'll remove "temporary". If anyone wants to write a BEP around it and make it part of the public interface, they are welcome :slightly_smiling_face:.